File: ChatClientVB/Services/ChatService.vb
Imports System.Net
Imports ChatClientVB
Imports Microsoft.AspNet.SignalR.Client

Public Class ChatService
    Implements IChatService

    Public Event PartecipanteDisconnesso(name As String) Implements IChatService.PartecipanteDisconnesso
    Public Event PartecipanteHaFattoAccesso(participant As Utente) Implements IChatService.PartecipanteHaFattoAccesso
    Public Event PartecipanteSiEDisconnesso(name As String) Implements IChatService.PartecipanteSiEDisconnesso
    Public Event PartecipanteRiconnesso(name As String) Implements IChatService.PartecipanteRiconnesso
    Public Event ConnessioneRiconnessione() Implements IChatService.ConnessioneRiconnessione
    Public Event ConnessioneRistabilita() Implements IChatService.ConnessioneRistabilita
    Public Event ConnessioneChiusa() Implements IChatService.ConnessioneChiusa
    Public Event NuovoMsgTesto(sender As String, msg As String, mt As TipoDiMessaggio) Implements IChatService.NuovoMsgTesto
    Public Event NuovoMsgImmagine(sender As String, img As Byte(), mt As TipoDiMessaggio) Implements IChatService.NuovoMsgImmagine
    Public Event PartecipanteStaScrivendo(name As String) Implements IChatService.PartecipanteStaScrivendo

    Private hubProxy As IHubProxy
    Private connessione As HubConnection
    Private url As String = "http://localhost:8080/InstantMessagingApp"

    Public Async Function ConnessioneAsync() As Task Implements IChatService.ConnessioneAsync
        connessione = New HubConnection(url)
        hubProxy = connessione.CreateHubProxy("ChatHub")
        hubProxy.On(Of Utente)("LoginPartecipante", Sub(u) RaiseEvent PartecipanteHaFattoAccesso(u))
        hubProxy.On(Of String)("LogoutPartecipante", Sub(n) RaiseEvent PartecipanteSiEDisconnesso(n))
        hubProxy.On(Of String)("DisconnessionePartecipante", Sub(n) RaiseEvent PartecipanteDisconnesso(n))
        hubProxy.On(Of String)("RiconnessionePartecipante", Sub(n) RaiseEvent PartecipanteRiconnesso(n))
        hubProxy.On(Of String, String)("MsgTestoBroadcast", Sub(n, m) RaiseEvent NuovoMsgTesto(n, m, TipoDiMessaggio.Broadcast))
        hubProxy.On(Of String, Byte())("MsgImmagineBroadcast", Sub(n, m) RaiseEvent NuovoMsgImmagine(n, m, TipoDiMessaggio.Broadcast))
        hubProxy.On(Of String, String)("MsgTestoUnicast", Sub(n, m) RaiseEvent NuovoMsgTesto(n, m, TipoDiMessaggio.Unicast))
        hubProxy.On(Of String, Byte())("MsgImmagineUnicast", Sub(n, m) RaiseEvent NuovoMsgImmagine(n, m, TipoDiMessaggio.Unicast))
        hubProxy.On(Of String)("PartecipanteStaScrivendo", Sub(p) RaiseEvent PartecipanteStaScrivendo(p))

        AddHandler connessione.Reconnecting, AddressOf Riconnessione
        AddHandler connessione.Reconnected, AddressOf Riconnesso
        AddHandler connessione.Closed, AddressOf Disconnesso

        ServicePointManager.DefaultConnectionLimit = 10
        Await connessione.Start()
    End Function

    Private Sub Disconnesso()
        RaiseEvent ConnessioneChiusa()
    End Sub

    Private Sub Riconnessione()
        RaiseEvent ConnessioneRiconnessione()
    End Sub

    Private Sub Riconnesso()
        RaiseEvent ConnessioneRistabilita()
    End Sub

    Public Async Function AccessoAsincr(nome As String, immagineProf As Byte()) As Task(Of List(Of Utente)) Implements IChatService.AccessoAsincr
        Dim utenti = Await hubProxy.Invoke(Of List(Of Utente))("Accesso", New Object() {nome, immagineProf})
        Return utenti
    End Function

    Public Async Function DisconnessioneAsincr() As Task Implements IChatService.DisconnessioneAsincr
        Await hubProxy.Invoke("Logout")
    End Function

    Public Async Function MandaMsgBroadcastAsincr(msg As String) As Task Implements IChatService.MandaMsgBroadcastAsincr
        Await hubProxy.Invoke("MsgTestoBroadcast", msg)
    End Function

    Public Async Function MandaMsgBroadcastAsincr(img As Byte()) As Task Implements IChatService.MandaMsgBroadcastAsincr
        Await hubProxy.Invoke("MsgImmagineBroadcast", img)
    End Function

    Public Async Function MandaMsgUnicastAsincr(destinatario As String, msg As String) As Task Implements IChatService.MandaMsgUnicastAsincr
        Await hubProxy.Invoke("MsgTestoUnicast", New Object() {destinatario, msg})
    End Function

    Public Async Function MandaMsgUnicastAsincr(destinatario As String, img As Byte()) As Task Implements IChatService.MandaMsgUnicastAsincr
        Await hubProxy.Invoke("MsgImmagineUnicast", New Object() {destinatario, img})
    End Function

    Public Async Function StaScrivendoAsincr(destinatario As String) As Task Implements IChatService.StaScrivendoAsincr
        Await hubProxy.Invoke("StaScrivendo", destinatario)
    End Function
End Class

File: ChatClientCS/Services/ChatService.cs
/*
Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. 
Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. 
Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. 
Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.


INSERISCI COMMENTO SPIEGAZIONE QUI



*/
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using ChatClientCS.Enums;
using ChatClientCS.Models;
using System.Net;
using Microsoft.AspNet.SignalR.Client;

namespace ChatClientCS.Services
{
    public class ChatService : IChatService
    {
        public event Action<string, string, TipoDiMessaggio> NuovoMsgTesto;
        public event Action<string, byte[], TipoDiMessaggio> NuovoMsgImmagine;
        public event Action<string> PartecipanteDisconnesso;
        public event Action<Utente> PartecipanteHaFattoAccesso;
        public event Action<string> PartecipanteSiEDisconnesso;
        public event Action<string> PartecipanteRiconnesso;
        public event Action RiconnessioneInCorso;
        public event Action ConnessioneRistabilita;
        public event Action ConnessioneChiusa;
        public event Action<string> PartecipanteStaScrivendo;

        private IHubProxy hubProxy;
        private HubConnection connessione;
        private string url = "http://localhost:8080/InstantMessagingApp";

        public async Task ConnessioneAsync()
        {
            connessione = new HubConnection(url);
            hubProxy = connessione.CreateHubProxy("ChatHub");
            hubProxy.On<Utente>("AccessoPartecipante", (u) => PartecipanteHaFattoAccesso?.Invoke(u));
            hubProxy.On<string>("LogoutPartecipante", (n) => PartecipanteSiEDisconnesso?.Invoke(n));
            hubProxy.On<string>("DisconnessionePartecipante", (n) => PartecipanteDisconnesso?.Invoke(n));
            hubProxy.On<string>("RiconnessionePartecipante", (n) => PartecipanteRiconnesso?.Invoke(n));
            hubProxy.On<string, string>("MsgTestoBroadcast", (n, m) => NuovoMsgTesto?.Invoke(n, m, TipoDiMessaggio.Broadcast));
            hubProxy.On<string, byte[]>("MsgImmagineBroadcast", (n, m) => NuovoMsgImmagine?.Invoke(n, m, TipoDiMessaggio.Broadcast));
            hubProxy.On<string, string>("MsgTestoUnicast", (n, m) => NuovoMsgTesto?.Invoke(n, m, TipoDiMessaggio.Unicast));
            hubProxy.On<string, byte[]>("MsgImmagineUnicast", (n, m) => NuovoMsgImmagine?.Invoke(n, m, TipoDiMessaggio.Unicast));
            hubProxy.On<string>("PartecipanteStaScrivendo", (p) => PartecipanteStaScrivendo?.Invoke(p));

            connessione.Reconnecting += Riconnessione;
            connessione.Reconnected += Riconnesso;
            connessione.Closed += Disconnesso;

            ServicePointManager.DefaultConnectionLimit = 10;
            await connessione.Start();
        }

        

        private void Riconnesso()
        {
            ConnessioneRistabilita?.Invoke();
        }
        private void Disconnesso()
        {
            ConnessioneChiusa?.Invoke();
        }
        private void Riconnessione()
        {
            RiconnessioneInCorso?.Invoke();
        }

        public async Task<List<Utente>> AccessoAsincr(string nome, byte[] immagineProf)
        {
            return await hubProxy.Invoke<List<Utente>>("Accesso", new object[] { nome, immagineProf });
        }

        public async Task DisconnessioneAsincr()
        {
            await hubProxy.Invoke("Logout");
        }

        public async Task MandaMsgBroadcastAsincr(string msg)
        {
            await hubProxy.Invoke("MsgTestoBroadcast", msg);
        }

        public async Task MandaMsgBroadcastAsincr(byte[] img)
        {
            await hubProxy.Invoke("MsgImmagineBroadcast", img);
        }

        public async Task MandaMsgUnicastAsincr(string destinatario, string msg)
        {
            await hubProxy.Invoke("MsgTestoUnicast", new object[] { destinatario, msg });
        }

        public async Task MandaMsgUnicastAsincr(string destinatario, byte[] img)
        {
            await hubProxy.Invoke("MsgImmagineUnicast", new object[] { destinatario, img });
        }

        public async Task StaScrivendoAsincr(string destinatario)
        {
            await hubProxy.Invoke("StaScrivendo", destinatario);
        }
    }
}